std/random.d

```diff
@@ -1336,6 +1330,271 @@ version (AnyARC4Random)
+        else version (AArch64) private enum NR_getrandom = 278;
+        else version (SPARC) private enum NR_getrandom = 347;
+        else version (SPARC64) private enum NR_getrandom = 347;
+        else version (Alpha) private enum NR_getrandom = 511;
+        else version (IA64) private enum NR_getrandom = 1339;
+        else version (PPC) private enum NR_getrandom = 359;
+        else version (PPC64) private enum NR_getrandom = 359;
+        else version (S390) private enum NR_getrandom = 349;
+        else version (SystemZ) private enum NR_getrandom = 349;
+        else version (HPPA) private enum NR_getrandom = 339;
+        else version (HPPA64) private enum NR_getrandom = 339;
+        else version (MIPS_O32) private enum NR_getrandom = 4353;
+        else version (MIPS_N32) private enum NR_getrandom = 6317;
+        else version (MIPS64) private enum NR_getrandom = 5313;
+        else version (SH)
+        {
+            static if (size_t.sizeof == 4)
+                private enum NR_getrandom = 373;
+            else static if (size_t.sizeof == 8)
+                private enum NR_getrandom = 384;
+        }
+
+        static if (is(typeof(NR_getrandom)))
+        {
+            // getrandom was introduced in Linux 3.17. If it's not available
+            // syscall will return -1 and errno will be set to ENOSYS.
+
+            /*
+             * Flags for getrandom(2)
+             *
+             * GRND_NONBLOCK    Don't block and return EAGAIN instead
+             * GRND_RANDOM      Use the /dev/random pool instead of /dev/urandom
+             */
+            private enum GRND_NONBLOCK = 0x0001;
+            private enum GRND_RANDOM = 0x0002;
+
+            /*
+                http://man7.org/linux/man-pages/man2/getrandom.2.html
+                If the urandom source has been initialized, reads of up to 256 bytes
+                will always return as many bytes as requested and will not be
+                interrupted by signals.  No such guarantees apply for larger buffer
+                sizes.
+            */
+            private extern(C) int syscall(size_t ident, size_t n, size_t arg1, size_t arg2) @nogc nothrow;
+        }
+    }
+
+    /+
+    Tries to use a system API to get entropy. Resources are not cached
+    due to the expectation that this function will not be called more
+    than once per program execution.
+
+    Params:
+        ptr = address at which to start writing
+        len = number of bytes to write
+
+    Returns:
+        Number of bytes successfully written up to `len`, or -1 on failure
+    +/
+    private ptrdiff_t oneTimeGenRandom()(scope void* ptr, size_t len) @system
+    {
+        version (linux)
+        static if (is(typeof(NR_getrandom)))
+        {{
+            import core.stdc.errno : ENOSYS, errno;
+            const result = syscall(NR_getrandom, cast(size_t) ptr, len, GRND_NONBLOCK);
+            if (result != -1 || errno != ENOSYS)
+                return result;
```

Comment:
> Do I understand right that this if block is to ensure that we will only return if the NR_getrandom syscall succeeds? And that if we fail, we drop through and use /dev/urandom instead?

Yes.

> If so, do I understand right that this is the expected behaviour on earlier linux kernels that don't support the syscall?

Yes. The earlier versions I reviewed/wrote used the C API to check before, but directly checking for `ENOSYS` is faster.

> However, since NR_getrandom was introduced in 3.17 (in 2014) and AFAICT all 3.x.y kernel releases are now EOL, this should not matter in practice.

Unfortunately it does. Some of my cheap VMs still run with 2.6.X because that's the latest release supported by OpenVZ and many hosters haven't switched to LXC. They even backported the Meltdown fixes to 2.6.X

> Finally: assuming the above, is it not possible to avoid the wasted syscall where the kernel does not support NR_getrandom? I assume not, because while one might be able to check at compile time whether the kernel supports this syscall, there's no guarantee that the compiled code will be running on top of the same kernel.

Yes, no compile-time guarantee.

Reply:
> Unfortunately it does. Some of my cheap VMs still run with 2.6.X because that's the latest release supported by OpenVZ and many hosters haven't switched to LXC.

Sure.  I meant that the performance cost of the guaranteed-to-fail syscall does not really matter, as these are really niche edge-cases for something that doesn't need to be _that_ fast in any case.